--- a/security/selinux/avc.c
+++ b/security/selinux/avc.c
@@ -700,28 +700,29 @@
 {
 	struct common_audit_data *ad = a;
 	struct selinux_audit_data *sad = ad->selinux_audit_data;
-	char buf[SELINUX_LABEL_LENGTH];
-	char *scontext = buf;
-	u32 scontext_len;
+	char *scontext, *tcontext;
+	u32 scontext_len, tcontext_len;
 	int rc;
+	const char *tclass;
 
 	trace_selinux_audited(sad);
 
-	rc = security_sid_to_context_stack(sad->ssid, &scontext,
+	rc = security_sid_to_context(sad->state, sad->ssid, &scontext,
 				     &scontext_len);
 	if (rc)
 		audit_log_format(ab, " ssid=%d", sad->ssid);
 	else
 		audit_log_format(ab, " scontext=%s", scontext);
 
-	rc = security_sid_to_context_stack(sad->tsid, &scontext,
-				     &scontext_len);
+	rc = security_sid_to_context(sad->state, sad->tsid, &tcontext,
+				     &tcontext_len);
 	if (rc)
 		audit_log_format(ab, " tsid=%d", sad->tsid);
 	else
-		audit_log_format(ab, " tcontext=%s", scontext);
+		audit_log_format(ab, " tcontext=%s", tcontext);
 
-	audit_log_format(ab, " tclass=%s", secclass_map[sad->tclass-1].name);
+	tclass = secclass_map[sad->tclass-1].name;
+	audit_log_format(ab, " tclass=%s", tclass);
 
 	if (sad->denied)
 		audit_log_format(ab, " permissive=%u", sad->result ? 0 : 1);
